File: algorithms/code/leetcode/lc006_zigzag_conversion/lc006_zigzag_conversion.py
#!/usr/bin/env python3
# https://leetcode.com/problems/zigzag-conversion/

import unittest


class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if s is None or numRows < 1:
            return ''
        buffer = ''
        n = len(s)
        scan = (numRows - 1) * 2
        scan = 1 if scan == 0 else scan
        for row in range(numRows):
            for i in range(row, n, scan):
                buffer += s[i]
                if 0 < row < numRows - 1:
                    offset = (numRows - 1 - row) * 2
                    if i + offset >= n:
                        break
                    buffer += s[i + offset]
        return buffer


class TestCode(unittest.TestCase):
    def test_1(self):
        self.assertEqual(
            Solution().convert('PAYPALISHIRING', 3), 'PAHNAPLSIIGYIR'
        )

    def test_2(self):
        self.assertEqual(
            Solution().convert('PAYPALISHIRING', 4), 'PINALSIGYAHRPI'
        )

    def test_3(self):
        self.assertEqual(Solution().convert('A', 1), 'A')

    def test_abcd(self):
        self.assertEqual(Solution().convert('ABCD', 3), 'ABDC')

    def test_abc(self):
        self.assertEqual(Solution().convert('ABC', 2), 'ACB')

    def test_nothing(self):
        self.assertEqual(Solution().convert(None, 2), '')

    def test_zero(self):
        self.assertEqual(Solution().convert('A', 0), '')

File: algorithms/code/leetcode/lc006_zigzag_conversion/lc006_zigzag_conversion.go
package lc006

// https://leetcode.com/problems/zigzag-conversion/

import (
	"strings"
)

func convert(s string, numRows int) string {
	if s == "" || numRows < 1 {
		return ""
	}
	// if numRows == 1 {
	// 	return s
	// }
	var buffer strings.Builder
	n := len(s)
	scan := (numRows - 1) * 2
	if scan == 0 {
		scan = 1
	}
	for row := 0; row < numRows; row++ {
		for i := row; i < n; i += scan {
			buffer.WriteByte(s[i])
			if row > 0 && row < numRows-1 {
				offset := (numRows - 1 - row) * 2
				if i+offset >= n {
					break
				}
				buffer.WriteByte(s[i+offset])
			}
		}
	}
	return buffer.String()
}
